File: python3/colorfight/building.py
from .constants import BLD_ENERGY_WELL, BLD_GOLD_MINE, BLD_FORTRESS
from .constants import BUILDING_COST, BUILDING_UPGRADE_COST, HOME_COST, HOME_UPGRADE_COST, DESTROY_BONUS

class BaseBuilding:
    cost = 0
    upgrade_cost = []
    def __init__(self):
        self.level = 1

    def get_energy_source(self, cell):
        return cell.energy

    def get_gold_source(self, cell):
        return cell.gold

    def get_attack_cost(self, cell):
        return cell.attack_cost 

    @property
    def is_empty(self):
        return self.name == 'empty'

    @property
    def is_home(self):
        return self.name == 'home'

    @property
    def max_level(self):
        if self.upgrade_cost:
            return len(self.upgrade_cost) + 1
        else:
            return 0

    @property
    def can_upgrade(self):
        return not self.is_empty and self.level < self.max_level

    @property
    def upgrade_gold(self):
        if self.can_upgrade:
            return self.upgrade_cost[self.level - 1][0]
        return None

    @property
    def upgrade_energy(self):
        if self.can_upgrade:
            return self.upgrade_cost[self.level - 1][1]
        return None

    @property
    def destroy_gold(self):
        if self.name == 'gold_mine':
            return DESTROY_BONUS[self.level - 1]
        return 0

    @property
    def destroy_forcefield(self):
        if self.name == 'energy_well':
            return DESTROY_BONUS[self.level - 1]
        return 0

    def info(self):
        return self.name

class Empty(BaseBuilding):
    name = 'empty'

class Home(BaseBuilding):
    name = 'home'
    cost = HOME_COST
    upgrade_cost = HOME_UPGRADE_COST

class EnergyWell(BaseBuilding):
    name = "energy_well"
    cost = BUILDING_COST
    upgrade_cost = BUILDING_UPGRADE_COST

class GoldMine(BaseBuilding):
    name = "gold_mine"
    cost = BUILDING_COST
    upgrade_cost = BUILDING_UPGRADE_COST

class Fortress(BaseBuilding):
    name = "fortress"
    cost = BUILDING_COST
    upgrade_cost = BUILDING_UPGRADE_COST

def get_building_class(building):
    '''
        return a class based on the string
    '''
    if building == BLD_ENERGY_WELL:
        return EnergyWell
    elif building == BLD_GOLD_MINE:
        return GoldMine
    elif building == BLD_FORTRESS:
        return Fortress
    else:
        return None

def letter_to_build_class(s):
    for cls in [Home, EnergyWell, GoldMine, Fortress]:
        if cls.name[0] == s:
            return cls
    return Empty

def str_to_build_class(s):
    for cls in [Empty, Home, EnergyWell, GoldMine, Fortress]:
        if cls.name == s:
            return cls
    return Empty

File: java/src/main/java/Building.java
import java.util.ArrayList;

abstract class Building {
    Pair cost = new Pair(0,0);
    String name;
    int level = 1;
    abstract boolean is_empty();
    abstract boolean is_home();

    ArrayList<Pair> upgrade_cost = new ArrayList<>();

    public Building(Pair cost, String name) {
        this.cost.assign(cost);
        this.name = name;
    }

    int max_level() {
        if(upgrade_cost.size()!=0){
            return upgrade_cost.size() + 1;
        }
        return 0;
    }

    boolean can_upgrade() {
        return level < max_level();
    }

    int upgrade_gold() {
        if(can_upgrade()){
            return upgrade_cost.get(level - 1).getX();
        }
        return Integer.parseInt(null);
    }

    int upgrade_energy() {
        if(can_upgrade()){
            return upgrade_cost.get(level - 1).getY();
        }
        return Integer.parseInt(null);
    }
}

class Empty extends Building {
    Empty(){
        super(new Pair(0,0),"empty");
    }

    @Override
    boolean is_empty() {
        return true;
    }

    @Override
    boolean is_home() {
        return false;
    }
}

class Home extends Building {
    Home(){
        super(new Pair(1000, 0),"home");
        upgrade_cost.add(new Pair(1000,1000));
        upgrade_cost.add(new Pair(2000,2000));
    }

    @Override
    boolean is_empty() {
        return false;
    }

    @Override
    boolean is_home() {
        return true;
    }
}

class EnergyWell extends Building {
    EnergyWell(){
        super(new Pair(100, 0),"energy_well");
        upgrade_cost.add(new Pair(200,0));
        upgrade_cost.add(new Pair(400,0));
    }

    @Override
    boolean is_empty() {
        return false;
    }

    @Override
    boolean is_home() {
        return false;
    }
}

class GoldMine extends Building {
    GoldMine(){
        super(new Pair(100, 0),"gold_mine");
        upgrade_cost.add(new Pair(200,0));
        upgrade_cost.add(new Pair(400,0));
    }

    @Override
    boolean is_empty() {
        return false;
    }

    @Override
    boolean is_home() {
        return false;
    }
}

class Fortress extends Building {
    Fortress(){
        super(new Pair(100, 0),"fortress");
        upgrade_cost.add(new Pair(200,0));
        upgrade_cost.add(new Pair(400,0));
    }

    @Override
    boolean is_empty() {
        return false;
    }

    @Override
    boolean is_home() {
        return false;
    }
}